tests/_utils/test_stages.py

```diff
@@ -0,0 +1,224 @@
+        t1 = datetime.now()
+
+        self.result = StageResult(procs, t1 - t0)
+
+    @property
+    def name(self) -> str:
+        return self.__class__.__name__
+
+    @property
+    def intro(self) -> str:
+        return banner(f"Entering stage: {self.name} tests") + "\n"
+
+    @property
+    def outro(self) -> str:
+        total = len(self.result.procs)
+        passed = len([p for p in self.result.procs if p.returncode == 0])
+
+        summary = bottom_line(self.name, total, passed)
+
+        footer = banner(
+            f"Exiting state: {self.name}",
+            details=(
+                "* Results      : "
+                + yellow(
+                    f"{passed} / {total} files passed ({passed/total*100:0.1f}%)"  # noqa E500
+                ),
+                "* Elapsed time : " + yellow(f"{self.result.time}"),
+            ),
+        )
+
+        return f"{summary}\n{footer}"
+
+    def run(
+        self, test_file: str, config: Config, system: System
+    ) -> CompletedProcess[str]:
+        ...
+
+    def file_args(self, test_file: str, config: Config) -> list[str]:
+        args = PER_FILE_ARGS.get(test_file, [])
+
+        # This is somewhat ugly but necessary in order to make pytest generate
+        # more verbose output for integration tests when --verbose is specified
+        if "integration" in test_file and config.verbose:
+            args += ["-v"]
+
+        return args
+
+    def _launch(
+        self, config: Config, system: System
+    ) -> list[CompletedProcess[str]]:
+
+        pool = multiprocessing.Pool(config.workers or 1)
+
+        jobs = [
+            pool.apply_async(self.run, (path, config, system))
+            for path in config.test_files
+        ]
+        pool.close()
+
+        return [job.get() for job in jobs]
+
+    def _log_proc(
+        self, proc: CompletedProcess[str], test_file: str, verbose: bool
+    ) -> None:
+        msg = f"({self.name}) {test_file}"
+        details = proc.stdout.split("\n") if verbose else None
+        if proc.returncode == 0:
+            LOG(passed(msg, details=details))
+        elif proc.returncode == SKIPPED_RETURNCODE:
+            LOG(skipped(msg))
+        else:
+            LOG(failed(msg, details=details))
+
+
+class CPU(TestStage):
+
+    kind = "cpus"
+
+    def run(
+        self, test_file: str, config: Config, system: System
+    ) -> CompletedProcess[str]:
+        test_path = join(config.root_dir, test_file)
+        stage_args = ["-cunumeric:test"] + next(self.cpu_args(config))
+        file_args = self.file_args(test_file, config)
+
+        cmd = [config.legate_path, test_path]
+        cmd += stage_args + file_args + config.extra_args
+
+        result = system.run(cmd, env=system.env)
+        self._log_proc(result, test_file, config.verbose)
+        return result
+
+    def cpu_args(self, config: Config) -> Iterator[list[str]]:
+        cpus = config.cpus or config.DEFAULT_CPUS
+        yield ["--cpus", str(cpus)]
```

Comment:
This generator approach is what will allow cycling through values for `--cpu-bind` for each test in the future. (Also for gpu below)